File: prep/icons.py
"""Inline SVG icon registry.

Phosphor Light icons (https://phosphoricons.com, MIT) are downloaded into
static/icons/ at design time, loaded into memory once at boot, and rendered
inline so they pick up `currentColor` and don't require an extra HTTP round
trip per use.

Usage from Jinja:
    {{ icon('arrow-left') }}
    {{ icon('check', class_='verdict-mark') }}

The svg blob already declares fill="currentColor" so any color/size styling
goes through the wrapper element. We strip the source's xmlns to keep the
serialized output compact (root <svg> in HTML5 doesn't need xmlns).
"""

from __future__ import annotations

from pathlib import Path

from markupsafe import Markup

# static/ lives at the repo root, not under the prep/ package.
_ICONS_DIR = Path(__file__).resolve().parent.parent / "static" / "icons"

# Loaded lazily on first call; cached process-wide. Editing an SVG on disk
# requires an app restart (same as templates/CSS in production).
_CACHE: dict[str, str] = {}


def _load(name: str) -> str:
    if name not in _CACHE:
        path = _ICONS_DIR / f"{name}.svg"
        if not path.exists():
            return ""  # missing icon: render nothing rather than break the page
        raw = path.read_text(encoding="utf-8").strip()
        _CACHE[name] = raw
    return _CACHE[name]


def icon(name: str, *, class_: str = "icon", title: str | None = None) -> Markup:
    """Render an inline SVG icon. Pass `class_` to scope styling; pass
    `title` to label decoratively-meaningless icons (most callers omit it
    and rely on aria-hidden via the wrapping element)."""
    svg = _load(name)
    if not svg:
        return Markup("")
    # Inject our class + aria-hidden right after the opening <svg.
    # Phosphor sources start with `<svg xmlns="..." viewBox="0 0 256 256" fill="currentColor">`
    # — keep their attributes intact, just add ours.
    open_end = svg.find(">")
    if open_end < 0:
        return Markup("")
    head = svg[:open_end]
    tail = svg[open_end:]
    extra = f' class="{class_}" aria-hidden="true"'
    if title:
        extra += f' role="img" aria-label="{title}"'
        # When titled the icon is no longer aria-hidden — replace.
        extra = extra.replace(' aria-hidden="true"', "")
    return Markup(head + extra + tail)

File: worker/runtime/adapters/nunjucks/icons.ts
// The `icon(name, class_=, title=)` global over the baked SVG map. The
// class and `aria-hidden` go right after the opening `<svg`, or `role="img"`
// and an `aria-label` when titled. An unknown name renders nothing.
import nunjucks from "nunjucks/browser/nunjucks-slim.js";

const { SafeString } = nunjucks.runtime;

interface IconKwargs {
  class_?: unknown;
  title?: unknown;
  __keywords?: true;
}

export function renderIcon(icons: Record<string, string>, name: string, class_ = "icon", title?: string): string {
  const svg = icons[name];
  if (!svg) return "";
  const openEnd = svg.indexOf(">");
  if (openEnd < 0) return "";
  const attrs = title
    ? ` class="${class_}" role="img" aria-label="${title}"`
    : ` class="${class_}" aria-hidden="true"`;
  return svg.slice(0, openEnd) + attrs + svg.slice(openEnd);
}

export function makeIconGlobal(icons: Record<string, string>) {
  return (name: unknown, kwargs?: IconKwargs): InstanceType<typeof SafeString> => {
    const kw = kwargs && typeof kwargs === "object" ? kwargs : {};
    const class_ = kw.class_ === undefined || kw.class_ === null ? "icon" : String(kw.class_);
    const title = kw.title ? String(kw.title) : undefined;
    return new SafeString(renderIcon(icons, String(name), class_, title));
  };
}
